### ISSUE TITLE
1.97 beta regression: ICE with "`Res::Err` but no error emitted"

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#157406]
This regression was discovered by the 1.97 beta crater run.

<details>
<summary>Error output</summary>

```
[INFO] [stdout] note: no errors encountered even though delayed bugs were created
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
[INFO] [stdout] note: those delayed bugs will now be shown as internal compiler errors
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] 
[INFO] [stdout] error: internal compiler error: `Res::Err` but no error emitted
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout]     |
[INFO] [stdout] note: delayed at /rustc-dev/96f70397ed7ba8baf8d1bb541a1ba8e06e3d6f57/compiler/rustc_hir_typeck/src/expr.rs:588:32
[INFO] [stdout]          0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
[INFO] [stdout]          1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
[INFO] [stdout]          2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
[INFO] [stdout]          3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
[INFO] [stdout]          4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_call
[INFO] [stdout]          6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
[INFO] [stdout]          8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          9: rustc_hir_typeck::check::check_fn
[INFO] [stdout]         10: rustc_hir_typeck::typeck_with_inspect::{closure#0}
[INFO] [stdout]         11: rustc_query_impl::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
[INFO] [stdout]         12: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, false>
[INFO] [stdout]         13: rustc_query_impl::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         14: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
[INFO] [stdout]         15: rustc_hir_analysis::check_crate
[INFO] [stdout]         16: rustc_interface::passes::analysis
[INFO] [stdout]         17: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, false>
[INFO] [stdout]         18: rustc_query_impl::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         19: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
[INFO] [stdout]         20: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
[INFO] [stdout]         21: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
[INFO] [stdout]         22: <std::sys::thread::unix::Thread>::new::thread_start
[INFO] [stdout]         23: <unknown>
[INFO] [stdout]         24: clone
[INFO] [stdout]       
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] note: rustc 1.97.0-beta.1 (96f70397e 2026-05-26) running on x86_64-unknown-linux-gnu
[INFO] [stdout] error: internal compiler error[E0618]: expected function, found `{type error}`
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^---------------
[INFO] [stdout]     |     |
[INFO] [stdout]     |     call expression requires function
[INFO] [stdout]     |
[INFO] [stdout] note: delayed at /rustc-dev/96f70397ed7ba8baf8d1bb541a1ba8e06e3d6f57/compiler/rustc_hir_typeck/src/callee.rs:747:34
[INFO] [stdout]          0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
[INFO] [stdout]          1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
[INFO] [stdout]          2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
[INFO] [stdout]          3: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_call
[INFO] [stdout]          4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
[INFO] [stdout]          6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          7: rustc_hir_typeck::check::check_fn
[INFO] [stdout]          8: rustc_hir_typeck::typeck_with_inspect::{closure#0}
[INFO] [stdout]          9: rustc_query_impl::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
[INFO] [stdout]         10: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, false>
[INFO] [stdout]         11: rustc_query_impl::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         12: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
[INFO] [stdout]         13: rustc_hir_analysis::check_crate
[INFO] [stdout]         14: rustc_interface::passes::analysis
[INFO] [stdout]         15: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, false>
[INFO] [stdout]         16: rustc_query_impl::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         17: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
[INFO] [stdout]         18: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
[INFO] [stdout]         19: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
[INFO] [stdout]         20: <std::sys::thread::unix::Thread>::new::thread_start
[INFO] [stdout]         21: <unknown>
[INFO] [stdout]         22: clone
[INFO] [stdout]       
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] 
[INFO] [stderr] note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2
[INFO] [stderr] 
[INFO] [stderr] note: some of the compiler flags provided by cargo are hidden
[INFO] [stderr] 
[INFO] [stderr] query stack during panic:
[INFO] [stderr] end of query stack
[INFO] [stderr] error: could not compile `api_xai`
```

</details>

https://crater-reports.s3.amazonaws.com/beta-1.97-3-v2/1.97.0-beta.1/reg/api_xai-0.3.0/log.txt

**Comment by theemathas** [comment]
Possibly another instance of the same issue in the **rustdoc** crater run:

```
[INFO] [stdout] note: no errors encountered even though delayed bugs were created
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] note: those delayed bugs will now be shown as internal compiler errors
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: internal compiler error: `Res::Err` but no error emitted
[INFO] [stdout]   --> src/command/mod.rs:42:14
[INFO] [stdout]    |
[INFO] [stdout] 42 |    .routine( command ::publish )
[INFO] [stdout]    |              ^^^^^^^^^^^^^^^^^
[INFO] [stdout]    |
[INFO] [stdout] note: delayed at /rustc-dev/96f70397ed7ba8baf8d1bb541a1ba8e06e3d6f57/compiler/rustc_hir_typeck/src/expr.rs:588:32
[INFO] [stdout]          0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
[INFO] [stdout]          1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
[INFO] [stdout]          2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
[INFO] [stdout]          3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
[INFO] [stdout]          4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args

(Full error is very long, and so is omitted for brevity.)
```

https://crater-reports.s3.amazonaws.com/beta-rustdoc-1.97-3/1.97.0-beta.1/reg/willbe-0.35.0/log.txt

**Comment by theemathas** [comment]
what on earth are these two crates (by the same author) doing with their module system...

**Comment by theemathas** [comment]
Smaller reproducer: [repro.zip](https://github.com/user-attachments/files/28657098/repro.zip)

src/lib.rs
```rust
use mod_interface::meta::mod_interface;

pub fn run() {
    publish();
}

mod private {
    pub fn publish() {}
}

crate::mod_interface! {
    layer publish;
    orphan use publish;
}
```

src/publish.rs
```rust
pub mod orphan {}
pub mod exposed {}
pub mod prelude {}
```

This is minimized from `willbe`, and depends on the crate `mod_interface`. I believe that the `api_xai` case is the same, as it also depends on `mod_interface`.

**Comment by theemathas** [comment]
Bisection result:

```
searched toolchains 7659cec4ed1457bc0d1f636127e66e8fe5008928 through 3bf5c6d99bc8a0c0d5b2f69826ed4f6d256a0a21


********************************************************************************
Regression in 17584a181979f04f2aaad867332c22db1caa511a
********************************************************************************

Attempting to search unrolled perf builds
Found commits ["7e7d5afe", "2de02f7e", "7407f925", "42270077", "308aaaf8", "5c6c8544", "39a07ed2", "7e319a2d", "75162d0a", "306c03c8", "a5fe24b5", "ff937bd0", "90b3efb7", "379bd301", "f68f2420", "3cd39685", "739f63e5", "4d336a43", "06d89fa1"]
installing 7e7d5afea1e05859a9df8ac528dd6a532e3c1468
uninstalling 7e7d5afea1e05859a9df8ac528dd6a532e3c1468
installing 2de02f7ee5a505c34a7befd6a84ebc25ae8e0517
uninstalling 2de02f7ee5a505c34a7befd6a84ebc25ae8e0517
installing 7407f9251f1188a9198c54124c9854b8839bbf82
uninstalling 7407f9251f1188a9198c54124c9854b8839bbf82
installing 42270077bd802aabdea052e977170cae9a02fb09
uninstalling 42270077bd802aabdea052e977170cae9a02fb09
installing 308aaaf85060a4545624535d43248fbb54d2c7b9
uninstalling 308aaaf85060a4545624535d43248fbb54d2c7b9
installing 5c6c8544fbff0b3b131b02ca39947698ff3027fc
uninstalling 5c6c8544fbff0b3b131b02ca39947698ff3027fc
installing 39a07ed20a4c5aaa55775f3f38cee787c8c86046
uninstalling 39a07ed20a4c5aaa55775f3f38cee787c8c86046
installing 7e319a2d3831b27b4d2aa1730416f791c87dcbc3
uninstalling 7e319a2d3831b27b4d2aa1730416f791c87dcbc3
installing 75162d0a0eee813a3af57c095460e95b69da7aa6
uninstalling 75162d0a0eee813a3af57c095460e95b69da7aa6
installing 306c03c8ad91f2f8ee82bba93de000ee6dcb6644
uninstalling 306c03c8ad91f2f8ee82bba93de000ee6dcb6644
installing a5fe24b570d7e00a24056873bd265fa290ac882d
uninstalling a5fe24b570d7e00a24056873bd265fa290ac882d
installing ff937bd07d15855ee19c7451d31f4c1d88a8c1bd
uninstalling ff937bd07d15855ee19c7451d31f4c1d88a8c1bd
installing 90b3efb7cc88ddd52da299f0d8d180dd7324e4a8
uninstalling 90b3efb7cc88ddd52da299f0d8d180dd7324e4a8
installing 379bd301ee8866a1af4f99cc4cf9b1ec8a9eb05d
uninstalling 379bd301ee8866a1af4f99cc4cf9b1ec8a9eb05d
installing f68f242073e6994183a543984ba0b744fd91922d
uninstalling f68f242073e6994183a543984ba0b744fd91922d
installing 3cd396850ee941cd8b35e85ef0e6ae3c5e5b5fe6
uninstalling 3cd396850ee941cd8b35e85ef0e6ae3c5e5b5fe6
installing 739f63e5ca79bcd2653a36da688e71d0322ac87b
uninstalling 739f63e5ca79bcd2653a36da688e71d0322ac87b
installing 4d336a430e12b9fa8885fc1a7fa831fa28c960d1
uninstalling 4d336a430e12b9fa8885fc1a7fa831fa28c960d1
installing 06d89fa160d8c23e75e42e6984e5e9a7e4fec2a8
uninstalling 06d89fa160d8c23e75e42e6984e5e9a7e4fec2a8
ERROR: none of the toolchains satisfied the predicate
```

Not sure what happened here with the bisection result. Maybe the ICE happens after the rollup, but the code compiles fine after the individual PRs?

The bisection points to the rollup in #155253, and the only PR in there that looks relevant is #154435. cc @petrochenkov @jackh726

**Comment by theemathas** [comment]
Minimized: [repro.zip](https://github.com/user-attachments/files/28657953/repro.zip)

src/lib.rs
```rust
use dep::my_macro;

pub fn foo() {
    ambiguous();
}

mod submodule {
    pub fn ambiguous() {}
}

pub mod ambiguous {}

my_macro!();
```

dep/src/lib.rs
```rust
#[proc_macro]
pub fn my_macro(_: proc_macro::TokenStream) -> proc_macro::TokenStream {
    r"
        use own::*;
        mod own {
            pub use super::submodule::*;
            pub use super::ambiguous;
        }
    "
    .parse()
    .unwrap()
}
```

<details>
<summary>Error output on beta</summary>

```
note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: `Res::Err` but no error emitted
 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^
  |
note: delayed at /rustc-dev/0417c25868d6dfbd1c291dfeae950504faa6f790/compiler/rustc_hir_typeck/src/expr.rs:588:32
         0: <std::backtrace::Backtrace>::create
         1: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         2: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         3: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         4: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
         5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_path
         6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
         8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
        10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
        11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_or_body_tail
        12: rustc_hir_typeck::check::check_fn
        13: rustc_hir_typeck::typeck_with_inspect::{closure#0}
        14: rustc_hir_typeck::typeck_root
        15: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, true>
        16: rustc_query_impl::query_impl::typeck_root::execute_query_incr::__rust_end_short_backtrace
        17: <rustc_middle::query::plumbing::TyCtxtEnsureOk>::typeck::<rustc_span::def_id::LocalDefId>
        18: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
        19: rustc_hir_analysis::check_crate
        20: rustc_interface::passes::analysis
        21: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, true>
        22: rustc_query_impl::query_impl::analysis::execute_query_incr::__rust_end_short_backtrace
        23: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
        24: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        25: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        26: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        27: <std::sys::thread::unix::Thread>::new::thread_start
        28: __pthread_cond_wait

 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^

error: internal compiler error[E0618]: expected function, found `{type error}`
 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^--
  |     |
  |     call expression requires function
  |
note: delayed at /rustc-dev/0417c25868d6dfbd1c291dfeae950504faa6f790/compiler/rustc_hir_typeck/src/callee.rs:747:34
         0: <std::backtrace::Backtrace>::create
         1: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         2: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         3: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::report_invalid_callee
         5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
         6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
         8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_or_body_tail
        10: rustc_hir_typeck::check::check_fn
        11: rustc_hir_typeck::typeck_with_inspect::{closure#0}
        12: rustc_hir_typeck::typeck_root
        13: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, true>
        14: rustc_query_impl::query_impl::typeck_root::execute_query_incr::__rust_end_short_backtrace
        15: <rustc_middle::query::plumbing::TyCtxtEnsureOk>::typeck::<rustc_span::def_id::LocalDefId>
        16: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
        17: rustc_hir_analysis::check_crate
        18: rustc_interface::passes::analysis
        19: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, true>
        20: rustc_query_impl::query_impl::analysis::execute_query_incr::__rust_end_short_backtrace
        21: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
        22: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        23: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        24: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        25: <std::sys::thread::unix::Thread>::new::thread_start
        26: __pthread_cond_wait

 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.97.0-beta.2 (0417c2586 2026-05-30) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `repro` (lib)
```

</details>

Bisection gave the same strange results as before.

**Comment by theemathas** [comment]
Given that the below code doesn't compile, on either stable or beta, maybe the above code shouldn't have compiled in the first place?

```rust
pub fn foo() {
    ambiguous();
}

mod submodule {
    pub fn ambiguous() {}
}

pub mod ambiguous {}

use own::*;
mod own {
    pub use super::submodule::*;
    pub use super::ambiguous;
}
```

**Comment by petrochenkov** [comment]
Fixed in https://github.com/rust-lang/rust/pull/157719 by reverting a part of https://github.com/rust-lang/rust/pull/154435.

@theemathas is right, that change uncovered some other issue and the code should likely produce an error, I'll investigate it later when I have time.

**Comment by theemathas** [comment]
Reopening to track beta backport.

**Comment by theemathas** [comment]
Fixed by #157787.
